crates/ironrdp-graphics/src/pointer.rs

```diff
@@ -0,0 +1,379 @@
+            }
+
+            self.current_byte = cursor.read_u8();
+        }
+
+        let bit = (self.current_byte >> (7 - self.read_bits)) & 1;
+        self.read_bits += 1;
+        bit
+    }
+}
+
+enum ColorStrideReader {
+    Color {
+        bpp: usize,
+        read_stide_bytes: usize,
+        stride_data_bytes: usize,
+        stride_padding: usize,
+    },
+    Bitmask(BitmaskStrideReader),
+}
+
+impl ColorStrideReader {
+    fn new(bpp: usize, stride: Stride) -> Self {
+        match bpp {
+            1 => Self::Bitmask(BitmaskStrideReader::new(stride)),
+            bpp => Self::Color {
+                bpp,
+                read_stide_bytes: 0,
+                stride_data_bytes: stride.data_bytes,
+                stride_padding: stride.padding,
+            },
+        }
+    }
+
+    fn next_pixel(&mut self, cursor: &mut ReadCursor<'_>) -> [u8; 4] {
+        match self {
+            ColorStrideReader::Color {
+                bpp,
+                read_stide_bytes,
+                stride_data_bytes,
+                stride_padding,
+            } => {
+                if read_stide_bytes == stride_data_bytes {
+                    *read_stide_bytes = 0;
+                    cursor.read_slice(*stride_padding);
+                }
+
+                match bpp {
+                    16 => {
+                        *read_stide_bytes += 2;
+                        let color_16bit = cursor.read_u16();
+                        rgb16_to_rgba32(color_16bit)
+                    }
+                    24 => {
+                        *read_stide_bytes += 3;
+
+                        let color_24bit = cursor.read_array::<3>();
+                        [color_24bit[2], color_24bit[1], color_24bit[0], 0xff]
+                    }
+                    32 => {
+                        *read_stide_bytes += 4;
+                        let color_32bit = cursor.read_array::<4>();
+                        [color_32bit[2], color_32bit[1], color_32bit[0], color_32bit[3]]
+                    }
+                    _ => panic!("BUG: should be validated in the calling code"),
+                }
+            }
+            ColorStrideReader::Bitmask(bitask) => {
+                if bitask.next_bit(cursor) == 1 {
+                    [0xff, 0xff, 0xff, 0xff]
+                } else {
+                    [0, 0, 0, 0xff]
+                }
+            }
+        }
+    }
+}
+
+fn bit_stride_size_align_u8(size_bits: usize) -> usize {
+    (size_bits + 7) / 8
+}
+
+fn bit_stride_size_align_u16(size_bits: usize) -> usize {
+    ((size_bits + 15) / 16) * 2
+}
+
+// TODO: Move to more general module
+fn rgb16_to_rgba32(color: u16) -> [u8; 4] {
+    let r = (((((color >> 11) & 0x1f) * 527) + 23) >> 6) as u8;
+    let g = (((((color >> 5) & 0x3f) * 259) + 33) >> 6) as u8;
+    let b = ((((color & 0x1f) * 527) + 23) >> 6) as u8;
+
+    [r, g, b, 0xff]
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
```

Comment:
Nit: do you think you could move this to the `ironrdp-testsuite-core` crate? Eventually [we want everything in there](https://github.com/Devolutions/IronRDP/blob/master/ARCHITECTURE.md#test-at-the-boundaries-test-features-not-code). Is there an okay way to trigger the same behavior by using the public API itself? Maybe we should add a feature flag to expose some inner API to the testsuite in some cases…